service/dap/server_test.go

```diff
@@ -2905,6 +2915,219 @@ func TestUnupportedCommandResponses(t *testing.T) {
 	})
 }
 
+// TestSetVariable tests SetVariable functionality that doesn't involve function calls.
+func TestSetVariable(t *testing.T) {
+	runTest(t, "testvariables", func(client *daptest.Client, fixture protest.Fixture) {
+		runDebugSessionWithBPs(t, client, "launch",
+			func() {
+				client.LaunchRequestWithArgs(map[string]interface{}{
+					"mode": "exec", "program": fixture.Path, "showGlobalVariables": true,
+				})
+			},
+			fixture.Source, nil,
+			[]onBreakpoint{{
+				execute: func() {
+					startLineno := 66 // after runtime.Breakpoint
+					if runtime.GOOS == "windows" && goversion.VersionAfterOrEqual(runtime.Version(), 1, 15) {
+						// Go1.15 on windows inserts a NOP after the call to
+						// runtime.Breakpoint and marks it same line as the
+						// runtime.Breakpoint call, making this flaky, so skip the line check.
+						startLineno = -1
+					}
+
+					handleStop(t, client, 1, "main.foobar", startLineno)
+
+					// We need 'args' to perform
+					//   VariableRequest -> SetVariableRequest -> VariableRequest.
+
+					client.VariablesRequest(1000)
+					args := client.ExpectVariablesResponse(t)
+
+					// Test: try to set composite type 'bar'. That is not implemented.
+					expectVarExact(t, args, 1, "bar", "bar", `main.FooBar {Baz: 10, Bur: "lorem"}`, hasChildren)
+					client.SetVariableRequest(1000, "bar", `main.FooBar {Baz: 42, Bur: "ipsum"}`)
+					if got, want := client.ExpectErrorResponse(t), "*ast.CompositeLit not implemented"; !strings.Contains(got.Body.Error.Format, want) {
+						t.Errorf("got %#v, want error string containing %q", got, want)
+					}
+
+					// Test: try to set integer type of composite type var 'bar'. That is supported.
```

Comment:
You should also be testing other composite types:
- pointers that have children with no names - do we handle that?
- arrays that have children with index names 
- maps that have different child structure depending on whether keys and values are scalars (3 subcases actually) and I think set variable might not make sense in some of them
- interfaces: can a user set .(data)?